consumer/consumererror/partialerror.go

```diff
@@ -66,3 +71,11 @@ func PartialMetricsError(err error, failedMetrics pdata.Metrics) error {
 func (err PartialError) GetMetrics() pdata.Metrics {
 	return err.failedMetrics
 }
+
+// IsPartial checks if an error was wrapped with a PartialError.
+func IsPartial(err error) bool {
+	if err != nil {
+		return errors.As(err, &PartialError{})
```

Comment:
Same question, do we trust the compiler here that no allocation is happening?

Reply:
Not sure, I'll see if I can craft a benchmark to test that.  My concern here had been that `errors.As` assigns `err` into the pointer it receives as a second parameter, which could result in trace/metric/log data being held onto if it's assigned into a package-scoped variable and never read from again.

Reply:
haha, also no thread safe :)) good point, benchmark is always for us to know, but we need this 🗡️

Reply:
It does seem to add an allocation and almost doubles the runtime, but it's still a matter of nanoseconds.

With pre-allocation:

```console
Running tool: /usr/local/bin/go test -benchmem -run=^$ -coverprofile=/var/folders/s4/yn176htd0y75yhy07spk_xgc0000gr/T/vscode-gooeO5Ae/go-code-cover -bench . go.opentelemetry.io/collector/consumer/consumererror

goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/consumer/consumererror
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkIsPartialNil-12      	579317856	         1.995 ns/op	       0 B/op	       0 allocs/op
BenchmarkIsPartialFalse-12    	19066316	        63.24 ns/op	       0 B/op	       0 allocs/op
BenchmarkIsPartialTrue-12     	18387475	        65.67 ns/op	       0 B/op	       0 allocs/op
PASS
coverage: 13.8% of statements
ok  	go.opentelemetry.io/collector/consumer/consumererror	4.577s
```

Safely:

```console
Running tool: /usr/local/bin/go test -benchmem -run=^$ -coverprofile=/var/folders/s4/yn176htd0y75yhy07spk_xgc0000gr/T/vscode-gooeO5Ae/go-code-cover -bench . go.opentelemetry.io/collector/consumer/consumererror

goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/consumer/consumererror
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkIsPartialNil-12      	595396323	         2.021 ns/op	       0 B/op	       0 allocs/op
BenchmarkIsPartialFalse-12    	10697533	       112.0 ns/op	      48 B/op	       1 allocs/op
BenchmarkIsPartialTrue-12     	10225606	       118.5 ns/op	      48 B/op	       1 allocs/op
PASS
coverage: 13.8% of statements
ok  	go.opentelemetry.io/collector/consumer/consumererror	4.721s
```